app/controllers/settings/two_factor_authentication/confirmations_controller.rb

```diff
@@ -18,18 +18,21 @@ def new
       end
 
       def create
-        if current_user.validate_and_consume_otp!(confirmation_params[:otp_attempt])
+        if current_user.validate_and_consume_otp!(confirmation_params[:otp_attempt], otp_secret: session[:new_otp_secret])
```

Comment:
What is this `new_otp_secret` from session?

Reply:
If you take a look at how the OTP setup flow works, basically when you click on the *Set Up*  button in the OTP setup page you are hitting `TwoFactorAuthentication::OtpAuthenticationController.rb#create` which generates a new `otp secret` that needs to be passed to this controller `TwoFactorAuthentication::ConfirmationsController.rb` in order to generate the QR Code – in the action `new` – and later to verify if the code 6-digit code introduced was valid – `create` action.